parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetWriter.java

```diff
@@ -403,15 +405,29 @@ public ParquetWriter(Path file, Configuration conf, WriteSupport<T> writeSupport
 
     this.codecFactory = codecFactory;
     CompressionCodecFactory.BytesInputCompressor compressor = codecFactory.getCompressor(compressionCodecName);
+
+    final String extraMetadataConfPrefix = ParquetOutputFormat.EXTRA_WRITE_METADATA + ".";
+    final Map<String, String> extraMetadata = new HashMap<>(writeContext.getExtraMetaData());
+
+    conf.iterator().forEachRemaining(entry -> {
```

Comment:
To support parquet writer without hadoop dependency, it would be good to implement this logic to `ParquetConfiguration` in two more overload constructors as well: https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetWriter.java#L341-L379

Reply:
👍 

Do you think it still makes sense to implement this functionality both as a new `ParquetWriter` constructor arg, *and* as a Configuration option that gets parsed during construction? Like if we add this new constructor overload:

```diff
ParquetWriter(
      OutputFile file,
      ParquetFileWriter.Mode mode,
      WriteSupport<T> writeSupport,
      CompressionCodecName compressionCodecName,
      long rowGroupSize,
      boolean validating,
      ParquetConfiguration conf,
      int maxPaddingSize,
      ParquetProperties encodingProps,
      FileEncryptionProperties encryptionProperties,
+     Map<String, String> extraMetadata)
```
It might be redundant/confusing to also iterate over `conf` for all keys starting with `parquet.write.metadata.extra`. Maybe we should just add it as a constructor arg and drop the Configuration key/parsing altogether. Wdyt?

Reply:
TBH, I don't think it is a good idea to add an overload of constructor (as we already have a bunch of constructors). What about adding a new method to ParquetWriter to set extra metadata? I have done similar things to parquet cpp: https://github.com/apache/arrow/commit/3bd57e35cc949bce6a04d7fbfed30c6db755b023

Reply:
yeah, that might be simplest! I could also add it to [ParquetProperties](https://github.com/apache/parquet-mr/blob/master/parquet-column/src/main/java/org/apache/parquet/column/ParquetProperties.java), since that's already included in the default `ParquetWriter` constructor.

Reply:
I think the `ParquetProperties` approach might be cleanest -- I prototyped adding an additional `ParquetWriter#addKeyValueMetadata` method (as in your Arrow PR), but the problem is that we'd have to re-initialize `ParquetWriter::writer` every time it's invoked, since it's constructed with a metadata argument: https://github.com/apache/parquet-mr/blob/238b9a2/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetWriter.java#L410

Reply:
Re-implemented it via `ParquetProperties` -- lmk what you think :)